lib/features/feed/cubit/feed_cubit.dart

```diff
@@ -10,17 +11,35 @@ part 'feed_state.dart';
 class FetchFeedCubit extends Cubit<FetchFeedState> {
   FetchFeedCubit({
     required ApiRepository apiRepository,
+    required DBRepository dBRepository,
   }) : super(const FetchFeedState.initial()) {
     _apiRepository = apiRepository;
+    _dBRepository = dBRepository;
   }
 
   late ApiRepository _apiRepository;
+  late DBRepository _dBRepository;
 
-  Future<void> fetchFeeds() async {
+  Future<void> fetchFeeds({
+    bool forceRefresh = false,
+  }) async {
     emit(const FetchFeedState.loading());
     try {
-      final feeds = await _apiRepository.fetchFeeds();
-      emit(FetchFeedState.loaded(fetchedFeed: feeds));
+      // Fetch feed entries from local database
+      final feedEntries = await _dBRepository.fetchLocalFeedEntries();
+      if (feedEntries.isNotEmpty && !forceRefresh) {
+        emit(FetchFeedState.loaded(fetchedFeed: feedEntries));
+        return;
+      }
+
+      // Fetch feed entries from API if local database is empty and persist them
+      if (feedEntries.isEmpty || forceRefresh) {
+        final feeds = await _apiRepository.fetchFeeds();
+        await _dBRepository.persistLocalFeedEntries(entries: feeds);
+        final feedEntries = await _dBRepository.fetchLocalFeedEntries();
+        emit(FetchFeedState.loaded(fetchedFeed: feedEntries));
+        return;
```

Comment:
Nice!

However, from this approach, it seems that the server is only hit when the entries are empty or the user forces a refresh. While this works, it seems that if new data is added and the user has some of the data locally on their device, the user may never see it. In my opinion, it takes intention from the user to call for a refresh, which I genuinely think will not be present from the users in this case. This is from a UX perspective.

To tackle this I suggest that after the `loaded` state is called, we do a fresh fetch to add new data in case there is. With this, the next time the user visits the page, they'll find the new data, without any extra effort.

Like so:

> 
>
>     if (feedEntries.isNotEmpty && !forceRefresh) {
>         emit(FetchFeedState.loaded(fetchedFeed: feedEntries));
> 
>        final feeds = await _apiRepository.fetchFeeds();
>        await _dBRepository.persistLocalFeedEntries(entries: feeds);
> 
>        return;
>      }
>
>
>

Reply:
I like your thinking. Last time I tried something similar, there was an issue with the cubits relaying data. I think this might fix it. Let me make a change and see what happens.

Reply:
Great! Let's see how it pans out

Reply:
So this works. Initially, I thought that we could have websockets deliver fresh data. But there might not be the capacity to implement that on the backend. Then having users do a hard refresh might prove a UX challenge. So this works as a measure in this scenario. Great catch.

Reply:
Great! Great!